File: frontend/src/shared/chat/tools/builtin/report.ts
// Report tools — let the LLM fetch generated daily/weekly/monthly reports and
// (with confirmation) trigger a fresh report run.
//
// 后端原语经注入的 ChatToolPlatform（electron loopback + 远程都走 HttpChatPlatform →
// serve-api /api/reports/* 端点，端点已就绪，本批仅前端接线）：
//   - report_list / report_get  → platform.listReports / getReport（silent，纯读）
//   - report_run                → platform.runReport（edit tier，同步跑 LLM 数十秒）
//
// createReportTools(platform) 闭包持 toolPlatform，由 createBuiltinTools 默认装入。

import type { ToolDef, ToolResult, ToolExecCtx } from '../registry'
import type { ChatToolPlatform } from '../../platform'

// ── helpers ────────────────────────────────────────────────────────────

function ok<O>(output: O, start: number): ToolResult<O> {
  return { ok: true, output, durationMs: Date.now() - start }
}

function err(code: string, message: string, start: number): ToolResult {
  return { ok: false, code, message, durationMs: Date.now() - start }
}

function asStr(x: unknown): string | undefined {
  return typeof x === 'string' && x.length > 0 ? x : undefined
}

function asInt(x: unknown, def: number, min: number, max: number): number {
  const n = typeof x === 'number' ? Math.floor(x) : NaN
  if (Number.isNaN(n)) return def
  return Math.min(Math.max(n, min), max)
}

type Cadence = 'daily' | 'weekly' | 'monthly'

function asCadence(x: unknown): Cadence | undefined {
  return x === 'daily' || x === 'weekly' || x === 'monthly' ? x : undefined
}

/** Resolve the effective input for an edit-tier tool (user may have edited the
 *  proposal in the ConfirmToolDialog). Mirrors write.ts::effective. */
function effective(input: unknown, ctx: ToolExecCtx): { resolved: unknown; userEdited: boolean } {
  if (ctx.userEditedInput !== undefined) {
    return { resolved: ctx.userEditedInput, userEdited: true }
  }
  return { resolved: input, userEdited: false }
}

/** Build the 3 report tools bound to the injected platform. */
export function createReportTools(platform: ChatToolPlatform): ToolDef[] {
  // ── 1. report_list — list generated reports ──────────────────────────────
  const reportList: ToolDef = {
    name: 'report_list',
    description:
      'List generated email reports (daily / weekly / monthly digests), newest first. ' +
      "Returns each report's id, cadence, report_date, status, headline, and summary counts " +
      '(total / unread / urgent / …) — NOT the full body. Use when the user asks "show my ' +
      'reports" / "what did the daily digest say" / "list this week\'s reports", then call ' +
      'report_get with an id to read one in full.',
    inputSchema: {
      type: 'object',
      properties: {
        cadence: {
          type: 'string',
          enum: ['daily', 'weekly', 'monthly'],
          description: 'Optional — restrict to one cadence. Omit for all.'
        },
        agent_id: {
          type: 'string',
          description: "Optional — restrict to one report agent's output."
        },
        limit: { type: 'integer', minimum: 1, maximum: 100, default: 20 }
      },
      required: []
    },
    confirmationTier: 'silent',
    category: 'read',
    surface: 'ipc',
    timeoutMs: 6000,
    handler: async (input, _ctx): Promise<ToolResult> => {
      const start = Date.now()
      const i = (input ?? {}) as Record<string, unknown>
      try {
        const items = await platform.listReports({
          cadence: asCadence(i.cadence),
          agentId: asStr(i.agent_id),
          limit: asInt(i.limit, 20, 1, 100)
        })
        return ok({ count: items.length, items }, start)
      } catch (e) {
        const code = (e as { code?: string }).code ?? 'E_INTERNAL'
        return err(code, e instanceof Error ? e.message : String(e), start)
      }
    }
  }

  // ── 2. report_get — full report detail (with doc blocks) ─────────────────
  const reportGet: ToolDef = {
    name: 'report_get',
    description:
      'Fetch a single report in full by its report_id (from report_list), including the ' +
      'structured block document (headline, sections, per-email items, key points, callouts). ' +
      'Use after report_list when the user wants the details of a specific report. ' +
      'Returns found:false if no report has that id.',
    inputSchema: {
      type: 'object',
      properties: {
        report_id: {
          type: 'string',
          description: 'The report id from a report_list result (e.g. "daily:daily:2026-06-01").'
        }
      },
      required: ['report_id']
    },
    confirmationTier: 'silent',
    category: 'read',
    surface: 'ipc',
    timeoutMs: 6000,
    handler: async (input, _ctx): Promise<ToolResult> => {
      const start = Date.now()
      const i = (input ?? {}) as Record<string, unknown>
      const id = asStr(i.report_id)
      if (!id) return err('E_INVALID_ARG', 'report_id is required (non-empty string)', start)
      try {
        const detail = await platform.getReport(id)
        if (!detail) return ok({ found: false, report_id: id }, start)
        return ok({ found: true, ...detail }, start)
      } catch (e) {
        const code = (e as { code?: string }).code ?? 'E_INTERNAL'
        return err(code, e instanceof Error ? e.message : String(e), start)
      }
    }
  }

  // ── 3. report_run — generate a fresh report now (runs the LLM) ───────────
  const reportRun: ToolDef = {
    name: 'report_run',
    description:
      'Generate a fresh report right now for the given report agent (runs the LLM over the ' +
      'agent\'s email window — can take tens of seconds). Use when the user asks to "run the ' +
      'daily report now" / "regenerate this week\'s digest". The user confirms before it runs ' +
      "(edit tier — they can change the agent_id / cadence). Returns the new report's id + " +
      'status + headline; call report_get with that id to read it.',
    inputSchema: {
      type: 'object',
      properties: {
        agent_id: {
          type: 'string',
          description: 'The report agent to run (its id from report_list / the /agents config).'
        },
        cadence: {
          type: 'string',
          enum: ['daily', 'weekly', 'monthly'],
          description:
            "Optional — override the agent's default cadence for this one run. Omit to use the " +
            "agent's configured cadence."
        }
      },
      required: ['agent_id']
    },
    confirmationTier: 'edit',
    category: 'write',
    surface: 'ipc',
    // Synchronous LLM run server-side — give it a generous budget (matches the
    // serve-api /run endpoint which awaits run_report_once).
    timeoutMs: 120_000,
    throttlePerMinute: 3,
    handler: async (input, ctx): Promise<ToolResult> => {
      const start = Date.now()
      const { resolved, userEdited } = effective(input, ctx)
      const i = (resolved ?? {}) as Record<string, unknown>
      const agentId = asStr(i.agent_id)
      if (!agentId) return err('E_INVALID_ARG', 'agent_id is required (non-empty string)', start)
      const cadence = asCadence(i.cadence)
      try {
        const result = await platform.runReport(agentId, cadence)
        return ok({ ...result, user_edited: userEdited }, start)
      } catch (e) {
        const code = (e as { code?: string }).code ?? 'E_INTERNAL'
        return err(code, e instanceof Error ? e.message : String(e), start)
      }
    }
  }

  return [reportList, reportGet, reportRun]
}

File: src/skills/builtin/report.py
"""report skill —— 跑 / 读 MailAgent 报告 Agent。

镜像 ``src/api/routers/reports.py``：``ReportStore`` + ``src/reports/wire.py`` 投影 +
``run_report_once``（in-process，**不 fork CLI**）。``report_run`` 是 P1 handoff 的执行能力，
返回 ``report_id`` 后可 ``report_get`` 取详情。
"""

from __future__ import annotations

import asyncio
import json
from typing import Any

from src.reports import wire
from src.skills.errors import SkillError
from src.skills.models import ToolDef, ToolHandler
from src.skills.registry import BoundSkill, BoundTool

_VALID_CADENCE = ("daily", "weekly", "monthly")


def _report_list(ctx: Any, params: dict[str, Any]) -> dict[str, Any]:
    store = ctx.report_store()
    rows = store.list_reports(
        cadence=params.get("cadence"),
        agent_id=params.get("agent_id"),
        limit=int(params.get("limit") or 50),
        offset=int(params.get("offset") or 0),
    )
    items = [wire.report_to_list_item(r) for r in rows]
    return {"items": items, "count": len(items)}


def _report_get(ctx: Any, params: dict[str, Any]) -> dict[str, Any]:
    report_id = str(params["report_id"])
    row = ctx.report_store().get_report(report_id)
    if row is None:
        raise SkillError("E_NOT_FOUND", f"report {report_id!r} not found", http_status=404)
    return wire.report_to_detail(row)


def _report_run(ctx: Any, params: dict[str, Any]) -> dict[str, Any]:
    """同步阻塞（blocking=True → invoke 跑在 to_thread）：独立 event loop 跑 run_report_once。"""
    from src.reports.worker import run_report_once

    store = ctx.report_store()
    agent_id = str(params["agent_id"])
    cadence = params.get("cadence")
    if cadence is not None and cadence not in _VALID_CADENCE:
        raise SkillError("E_INVALID_ARG", f"cadence must be {'|'.join(_VALID_CADENCE)}")

    agent = store.get_agent(agent_id)
    if agent is None:
        raise SkillError("E_NOT_FOUND", f"report_agent {agent_id!r} not found", http_status=404)
    if agent.get("type", "report") != "report":
        raise SkillError(
            "E_INVALID_ARG",
            f"agent {agent_id!r} is type {agent.get('type')!r}, not a report agent; "
            "manual run is report-only",
        )
    if cadence is not None:
        try:
            sched = json.loads(agent.get("schedule_json") or "{}") or {}
        except (json.JSONDecodeError, TypeError):
            sched = {}
        sched["cadence"] = cadence
        agent = {**agent, "schedule_json": json.dumps(sched, ensure_ascii=False)}

    try:
        rid = asyncio.run(run_report_once(store=store, db_path=store.db_path, agent=agent))
    except Exception as exc:  # noqa: BLE001 — 兜底成结构化 skill error
        raise SkillError("E_LLM_FAILED", f"report generation failed: {exc}", http_status=500)

    row = store.get_report(rid) or {}
    return {
        "report_id": rid,
        "status": row.get("status", "unknown"),
        "headline": row.get("headline") or "",
        "cadence": row.get("cadence"),
        "report_date": row.get("report_date"),
        "error": row.get("error"),
    }


def build_skill() -> BoundSkill:
    tools = [
        BoundTool(
            ToolDef(
                name="report_list",
                description="List generated reports (newest first; no block bodies).",
                input_schema={
                    "type": "object",
                    "properties": {
                        "cadence": {"type": "string", "enum": list(_VALID_CADENCE)},
                        "agent_id": {"type": "string"},
                        "limit": {"type": "integer"},
                        "offset": {"type": "integer"},
                    },
                },
                output_schema={"type": "object", "description": "{items, count}"},
                confirmation_tier="none",
                side_effect="read",
                auth_scopes=["report:read"],
                mcp_exposed=True,
                handler=ToolHandler(kind="service", target="ReportStore.list_reports"),
            ),
            _report_list,
        ),
        BoundTool(
            ToolDef(
                name="report_get",
                description="Fetch one report's full detail (blocks doc + counts).",
                input_schema={
                    "type": "object",
                    "properties": {"report_id": {"type": "string"}},
                    "required": ["report_id"],
                },
                output_schema={"type": "object"},
                confirmation_tier="none",
                side_effect="read",
                auth_scopes=["report:read"],
                mcp_exposed=True,
                handler=ToolHandler(kind="service", target="ReportStore.get_report"),
            ),
            _report_get,
        ),
        BoundTool(
            ToolDef(
                name="report_run",
                description="Run a report agent now (calls the LLM); returns the new report_id.",
                input_schema={
                    "type": "object",
                    "properties": {
                        "agent_id": {"type": "string"},
                        "cadence": {"type": "string", "enum": list(_VALID_CADENCE)},
                    },
                    "required": ["agent_id"],
                },
                output_schema={
                    "type": "object",
                    "description": "{report_id, status, headline, cadence, report_date, error}",
                },
                confirmation_tier="preview",
                side_effect="external_call",
                auth_scopes=["report:run"],
                mcp_exposed=True,
                timeout_ms=120000,
                handler=ToolHandler(kind="service", target="reports.run_report_once"),
            ),
            _report_run,
            blocking=True,
        ),
    ]
    return BoundSkill(
        name="report",
        version="1.0.0",
        title="Reports",
        description="Run and read MailAgent report agents (daily/weekly/monthly digests).",
        default_enabled=True,
        prompt_fragment=(
            "Use report_list to see existing reports, report_get to read one, and report_run "
            "to generate a fresh report for an agent_id (this calls the LLM and returns a "
            "report_id you can then report_get)."
        ),
        docs_path="skills/report/SKILL.md",
        tools=tools,
    )
